### ISSUE TITLE
App freezes after 64th call to findEvent in loop

### ISSUE THREAD

**Issue by tcowin** [EddyVerbruggen/Calendar-PhoneGap-Plugin#234]
I have one page in my app that displays a list of events - it has 77 events now. By each event, I offer a link to 'add to calendar' or 'remove from calendar', depending upon if the event is in the Calendar. After populating the primary html of the list(without those add/remove links), I go through another loop to call findEvent on each and then add the appropriate link to each event listing via javascript. This has been working fine over the last few years, but when I try it in a new app with the very latest _everything_, it freezes on the 64th call without an error. If I limit it to 63 calls, then it works perfectly. But once I do the 64th call, none of the success callback functions fire. It seems to be the same behavior on an iPhone 6 or the emulated iPhone. Is there some sort of call stack I could be over running? It appears to be using 42MB when it freezes, almost as if it has a limit of 42 on the real iPhone, but in emulation it uses more memory. I get no errors...

**Comment by tcowin** [comment]
I've tried this with a bare bones, stripped down call to eventsMatchingPredicate:
(but got the same behavior)

```ObjectiveC

- (NSArray*) findEKEventsWithTitleAndDate: (NSString *)title
                       startDate: (NSDate *)startDate
                         endDate: (NSDate *)endDate {

  if (title != (id)[NSNull null] && title.length > 0) {
    title = [title stringByReplacingOccurrencesOfString:@"'" withString:@"\\'"];
  }

  NSString *predicateString = [NSString stringWithFormat:@"title contains[c] '%@'", title];

  NSPredicate *matches;
  NSArray  *datedEvents, *matchingEvents;

  if (predicateString.length > 0) {
    matches = [NSPredicate predicateWithFormat:predicateString];
 
    datedEvents = [self.eventStore eventsMatchingPredicate:[eventStore predicateForEventsWithStartDate:startDate endDate:endDate calendars:nil]];

    matchingEvents = [datedEvents filteredArrayUsingPredicate:matches];
  } else {
 
    datedEvents = [self.eventStore eventsMatchingPredicate:[eventStore predicateForEventsWithStartDate:startDate endDate:endDate calendars:nil]];

    matchingEvents = datedEvents;
  }

  return matchingEvents;
}
```

**Comment by ptmahoney** [comment]
I think the problem I'm having is related to what you have found. So thought it best to add the details of my issue here rather then open a new one. I hope that's OK :^)

I too am rendering a list. For each item on the list I'm calling  <i>findEvent()</i> with a separate success callback function for each. I'm randomly seeing some of my callbacks called multiple times and, consequently, others not called at all.

Having scratched my head for some days now, I put some <i>console.log()</i> calls in the code. This is the output:
<hr>
<b>item 1 find request</b>

RQ: CallbackId = Calendar534918123 cordova.js:937

RQ: argsJson = [{"title":"Book Highwoods Competition","location":"Highwoods Golf Club","notes":"1st Tee - Men and Ladies Stableford £3","startTime":1448046000000,"endTime":1448047800000,"options":{"firstReminderMinutes":60,"secondReminderMinutes":null,"recurrence":null,"recurrenceInterval":1,"recurrenceEndDate":null,"calendarName":null,"calendarId":null,"url":null}}] cordova.js:938

<b>item 2 find request</b>

RQ: CallbackId = Calendar534918124 cordova.js:937

RQ: argsJson = [{"title":"Book Highwoods Competition","location":"Highwoods Golf Club","notes":"1st Tee - Men and Ladies Stableford £3","startTime":1448650800000,"endTime":1448652600000,"options":{"firstReminderMinutes":60,"secondReminderMinutes":null,"recurrence":null,"recurrenceInterval":1,"recurrenceEndDate":null,"calendarName":null,"calendarId":null,"url":null}}] cordova.js:938

<b>item 3 find request</b>

RQ: CallbackId = Calendar534918125 cordova.js:937

RQ: argsJson = [{"title":"Book Highwoods Competition","location":"Highwoods Golf Club","notes":"1st Tee - Ladies Stableford £2","startTime":1448910000000,"endTime":1448911800000,"options":{"firstReminderMinutes":60,"secondReminderMinutes":null,"recurrence":null,"recurrenceInterval":1,"recurrenceEndDate":null,"calendarName":null,"calendarId":null,"url":null}}] cordova.js:938

<b>item 4 find request</b>

RQ: CallbackId = Calendar534918126 cordova.js:937

RQ: argsJson = [{"title":"Book Highwoods Competition","location":"Highwoods Golf Club","notes":"1st Tee - Mens Xmas Pudding Stableford £3","startTime":1448996400000,"endTime":1448998200000,"options":{"firstReminderMinutes":60,"secondReminderMinutes":null,"recurrence":null,"recurrenceInterval":1,"recurrenceEndDate":null,"calendarName":null,"calendarId":null,"url":null}}] cordova.js:938

<b>item 5 find request</b>

RQ: CallbackId = Calendar534918127 cordova.js:937

RQ: argsJson = [{"title":"Book Highwoods Competition","location":"Highwoods Golf Club","notes":"1st Tee - Friends and Family Morning","startTime":1449860400000,"endTime":1449862200000,"options":{"firstReminderMinutes":60,"secondReminderMinutes":null,"recurrence":null,"recurrenceInterval":1,"recurrenceEndDate":null,"calendarName":null,"calendarId":null,"url":null}}] cordova.js:938

<b>response message 1</b>

RQ: Message = S11 Calendar534918123 [] - RQ: Keep callback = true cordova.js:1080

<b>response message 2</b>

RQ: Message = S11 Calendar534918127 [] - RQ: Keep callback = true cordova.js:1080

<b>response message 3</b>

RQ: Message = S11 Calendar534918127 [] - RQ: Keep callback = true 

<b>response message 4</b>

RQ: Message = S11 Calendar534918127 [] - RQ: Keep callback = true 

<b>response message 5</b>

RQ: Message = S11 Calendar534918127 [] - RQ: Keep callback = true 
<hr>

It varies from run to run. But if you study the above: callback ID 534918123 is called once; callback ID 534918127 is called 4 times. The other 3 callbacks are not called at all.

The other thing to note is the <b>S1</b> part of the response message. S = success, which is good. But the '1' is used to set a flag call <i>keepCallback</i>. The callbacks passed to cordova are stored in a hash called <i>cordova.callbacks</i>. As I've refreshed my list a number of times while trying to debug this issue, there are now 175 callbacks in the hash!

Struggling to see where I've gone wrong here. Summary is:

1. Why is the Calendar plugin responding with the wrong callback ID 
2. Why is the Calendar plugin saying keep the callback.

P.S As there are no matching events in my calender, the response is always an empty array. []

**Comment by tcowin** [comment]
My general impression is that the iOS implementation here may not be thread safe for concurrent use - so that if you are making these calls in a loop, you'll continue getting these unpredictable results.

For this problem: https://github.com/EddyVerbruggen/Calendar-PhoneGap-Plugin/issues/239 I took the code to createEvent out from behind the 'runInBackground' call in Calendar.m, and it all started working correctly. I don't know that I tried that with my findEvent calls, but I suspect that might work.
(not sure if you're comfortable modifying the Cocoa iOS code?)

I took a different approach with this findEvent loop - since all the events that I'm looking for are in a distinct time period, I just do one findEvent call with the only conditions being the date range(you should be able to call findEvent with title, location, notes all set to null), and then store all the events in a js array and then deal with them locally there.

Hope this helps... let me know what you find...

**Comment by ptmahoney** [comment]
Thank you for you reply Tom. I like the idea of doing one query to get all event in one go into a local cache. I forgot to say that in my debugging session I was actually running against an Android emulation. Unless Eddy is available to look at this, which is quite understandable, I think I will now dig deeper into the Java code and see what I find.

**Comment by EddyVerbruggen** [comment]
(This response is not meant to be defensive at all, just wanting to shed some light on a few design decisions I made in the past)

You guys are right about the non thread safeness. This plugin was not designed with tackling Cordova's async nature in mind, so it expects you to not call the same method in parallel but rather when its callback is invoked.

Note that people tend to get scared when Cordova logs about a plugin not running in a background thread but taking a long time to complete. Which is a valid complaint, but not in all circumstances. So that's why most stuff is in a background thread.

Keep the discussion going, and if anyone thinks he has a great fix for an issue please open a PR. I'm always looking for folks wanting to contribute, so if you want to be a collaborator on the repo just say so. As long as you send changes as a PR before merging them and try not to break backward compatibility I'm good.

**Comment by ptmahoney** [comment]
I have some ideas to try out on Calendar.java - so will take the plunge. If that works out, I'll repeat for Calendar.m

The GitHub and Pull Request stuff is foreign to me. So any pointers would be great.

This is definitely a background task for me to please don't hold you breath :^)

**Comment by EddyVerbruggen** [comment]
Awesome @ptmahoney, I hope you can find some time. No rush of course.

You can hit that 'fork' button at the top-right and it will clone this repo to your account. Then 'git clone https://github.com/ptmahoney/Calendar-PhoneGap-Plugin' on your local machine and adjust any code you like. Commit and push once done. Then go to 'https://github.com/ptmahoney/Calendar-PhoneGap-Plugin/pulls' and hit the 'New Pull Request' button.

The missing piece here is a Git client but that totally depends on the tools you're used to so I'm not going to advise any.

**Comment by tcowin** [comment]
I’m happy to help as well - same deal here though as I’m a bit backed up…

Thanks,

Tom
--
tom.cowin@blackfin.biz
Blackfin Software, LLC
(425)985-3150

**Comment by EddyVerbruggen** [comment]
Thanks guys, much appreciated. Take your time. I know I will ;)

**Comment by ptmahoney** [comment]
I have managed to resolve my concurrency issue with Android. Please see PR https://github.com/EddyVerbruggen/Calendar-PhoneGap-Plugin/pull/253

I have a few other things to code before I see if I have the same problem with iOS. But wanted to see if someone would sanity check what I've done for Android first.

**Comment by ptmahoney** [comment]
I have found that the iOS code does not suffer from the concurrent/re-entrant access issue that the Android code had.
